src/features/weak-references.md

```diff
@@ -54,104 +54,133 @@ A JavaScript `WeakMap` is not really _weak_: it actually refers _strongly_ to it
+      const deref = weakRef.deref(); // 2
+      if (deref !== undefined) deref(ev); // 3
+    });
+    socket.addEventListener("message", listenerWrapper);
+    gListenersRegistry.register(listener,
+                                { socket, listenerWrapper }); // 4
   }
-});
-
-function getImageCached(name) {
-  const ref = cache.get(name); // 1
-  if (ref !== undefined) { // 2
-    const deref = ref.deref();
-    if (deref !== undefined) return deref;
-  }
-  const image = performExpensiveOperation(name); // 3
-  const wr = new WeakRef(image); // 4
-  cache.set(name, wr); // 5
-  finalizationRegistry.register(image, name); // 6
-  return image; // 7
 }
 ```
 
-Given an image name, we look up its corresponding weak reference in the cache (1). If the weak reference still points to something (2), we can return the cached image data. If there’s no cache entry yet for this image name, or if the cached image data has been garbage-collected, we compute the image data (3), create a new weak reference to it (4), store the image name and the weak reference in the cache (5), register a finalizer that removes the image name from the cache once the image data is garbage-collected (6), and return the image (7).
+We wrap the event listener that does the work in a `WeakRef` to make garbage-collectible, and to not leak its reference to the `MovingAvg` instance via `this` (1). We make a wrapper that `deref` the `WeakRef` to check if it is still alive (2), then call it if so (3). After registering the listener wrapper on the socket, we also register the inner listener on the `FinalizationRegistry`, passing a _holding value_ `{ socket, wrapper }` to the registration (4). Sometime after the inner listener garbage-collected, the finalizer may run, with the holding value passed to it. Inside the finalizer, we unregister the listener wrapper as well, making all memory associated with a `MovingAvg` instance garbage-collectible (5).
+
+With all this, our original implementation of `MovingAvgComponent` neither leaks memory nor requires any manual disposal.
 
 ## Don’t overdo it { #progressive-enhancement }
 
 After hearing about these new capabilities, it might be tempting to `WeakRef` All The Things™. However, that’s probably not a good idea. Some things are explicitly _not_ good use cases for `WeakRef`s and finalizers.
 
-In general, avoid writing code that depends on the garbage collector cleaning up a `WeakRef` or calling a finalizer at any predictable time — [it can’t be done](https://github.com/tc39/proposal-weakrefs#a-note-of-caution)!
+In general, avoid writing code that depends on the garbage collector cleaning up a `WeakRef` or calling a finalizer at any predictable time — [it can’t be done](https://github.com/tc39/proposal-weakrefs#a-note-of-caution)! Moreover, whether an object is garbage-collectible at all may depend on implementation details, such as the representation of closures, that are both subtle and may differ between different versions of V8.
```

Comment:
Let’s not just talk about V8 here but also mention different JavaScript engines. Maybe something like:

```suggestion
In general, avoid writing code that depends on the garbage collector cleaning up a `WeakRef` or calling a finalizer at any predictable time — [it can’t be done](https://github.com/tc39/proposal-weakrefs#a-note-of-caution)! Moreover, whether an object is garbage-collectible at all may depend on implementation details, such as the representation of closures, that are both subtle and may differ between different versions of V8 and across JavaScript engines.
```

Reply:
Adopted.